Mathlib/Analysis/SpecialFunctions/Elliptic/Weierstrass.lean

```diff
@@ -281,6 +299,280 @@ lemma weierstrassP_neg (z : ℂ) : ℘[L] (-z) = ℘[L] z := by
+  simp only [Equiv.neg_apply, NegMemClass.coe_neg, sub_neg_eq_add, neg_add_eq_sub,
+    ← div_neg, ← tsum_neg, apply_ite, neg_zero]
+  congr! 3 with l
+  · simp [neg_eq_iff_eq_neg]
+  ring
+
+end derivWeierstrassPExcept
+
+section Periodicity
+
+lemma derivWeierstrassPExcept_add_coe (l₀ : ℂ) (z : ℂ) (l : L.lattice) :
+    ℘'[L - l₀] (z + l) = ℘'[L - (l₀ - l)] z := by
+  simp only [derivWeierstrassPExcept]
+  rw [← (Equiv.addRight l).tsum_eq]
+  simp only [Equiv.coe_addRight, Submodule.coe_add, add_sub_add_right_eq_sub, eq_sub_iff_add_eq]
+
+-- Subsumed by `weierstrassP_add_coe`
+private lemma weierstrassPExcept_add_coe_aux
+    (l₀ : ℂ) (hl₀ : l₀ ∈ L.lattice) (l : L.lattice) (hl : l.1 / 2 ∉ L.lattice) :
+    Set.EqOn (℘[L - l₀] <| · + l) (℘[L - (l₀ - l)] · + (1 / l₀ ^ 2 - 1 / (l₀ - ↑l) ^ 2))
+      (L.lattice \ {l₀ - l})ᶜ := by
+  apply IsOpen.eqOn_of_deriv_eq (𝕜 := ℂ) L.isOpen_compl_lattice_diff
+    ?_ ?_ ?_ ?_ (x := - (l / 2)) ?_ ?_
+  · refine (Set.Countable.isConnected_compl_of_one_lt_rank (by simp) ?_).2
+    exact .mono sdiff_le (countable_of_Lindelof_of_discrete (X := L.lattice))
+  · refine (L.differentiableOn_weierstrassPExcept l₀).comp (f := (· + l.1)) (by fun_prop) ?_
+    rintro x h₁ ⟨h₂ : x + l ∈ _, h₃ : x + l ≠ l₀⟩
+    exact h₁ ⟨by simpa using sub_mem h₂ l.2, by rintro rfl; simp at h₃⟩
+  · refine .add (L.differentiableOn_weierstrassPExcept _) (by simp)
+  · intro x hx
+    simp only [deriv_add_const', deriv_comp_add_const]
+    rw [L.eqOn_deriv_weierstrassPExcept_derivWeierstrassPExcept,
+      L.eqOn_deriv_weierstrassPExcept_derivWeierstrassPExcept, L.derivWeierstrassPExcept_add_coe]
+    · simpa using hx
+    · simp only [Set.mem_compl_iff, Set.mem_diff, SetLike.mem_coe, Set.mem_singleton_iff, not_and,
+        Decidable.not_not, eq_sub_iff_add_eq] at hx ⊢
+      exact fun H ↦ hx (by simpa using sub_mem H l.2)
+  · simp [hl]
+  · rw [L.weierstrassPExcept_neg, L.weierstrassPExcept_def ⟨l₀, hl₀⟩,
+      L.weierstrassPExcept_def ⟨_, neg_mem (sub_mem hl₀ l.2)⟩, add_assoc]
+    congr 2 <;> ring
+
+-- Subsumed by `weierstrassP_add_coe`
+private lemma weierstrassP_add_coe_aux (z : ℂ) (l : L.lattice) (hl : l.1 / 2 ∉ L.lattice) :
+    ℘[L] (z + l) = ℘[L] z := by
+  have hl0 : l ≠ 0 := by rintro rfl; simp at hl
+  by_cases hz : z ∈ L.lattice
+  · have := L.weierstrassPExcept_add_coe_aux (z + l) (add_mem hz l.2) l hl (x := z) (by simp)
+    dsimp at this
+    rw [← L.weierstrassPExcept_add ⟨z + l, add_mem hz l.2⟩, this,
+      ← L.weierstrassPExcept_add ⟨z, hz⟩]
+    simp
+    ring
+  · have := L.weierstrassPExcept_add_coe_aux 0 (zero_mem _) l hl (x := z) (by simp [hz])
+    simp only [zero_sub, ne_eq, OfNat.ofNat_ne_zero, not_false_eq_true, zero_pow, div_zero,
+      even_two, Even.neg_pow, one_div] at this
+    rw [← L.weierstrassPExcept_add 0, Submodule.coe_zero, this, ← L.weierstrassPExcept_add (-l)]
+    simp
+    ring
+
+@[simp]
+lemma weierstrassP_add_coe (z : ℂ) (l : L.lattice) : ℘[L] (z + l) = ℘[L] z := by
```

Comment:
It might be worthwhile to connect this up to `Function.Periodic` too, for each of the periods in `L`, to give access to that API.